bus.py

```diff
@@ -0,0 +1,106 @@
+import random
+from utility import distance, firstn
+
+
+def _road_point_generator(roads):
+    assert len(roads) > 0
+
+    # Length of all roads combined
+    total_length = sum(map(lambda road: road.getLength(), roads))
+
+    while True:
+        # Select a point on the combined stretch of road
+        distance = random.uniform(0, total_length)
+
+        # Find the selected road
+        length_sum = 0.0
+        found_road = False
+        for road in roads:
+            from_node = road.getFromNode().getCoord()
+            to_node = road.getToNode().getCoord()
+
+            if (length_sum + road.getLength()) >= distance:
+                # Compute the exact point on the selected road
+
+                # Distance along the road segment
+                remaining = length_sum - distance
+
+                # Vector for the road segment
+                vector = (
+                        to_node[0] - from_node[0],
+                        to_node[1] - from_node[1])
+
+                # Normalized vector used for direction
+                direction = (
+                        vector[0]/road.getLength(),
+                        vector[1]/road.getLength())
+
+                yield [
+                        from_node[0] + remaining * direction[0],
+                        from_node[1] + remaining * direction[1],
+                        road,
+                        remaining]
+                found_road = True
+                break
+            else:
+                length_sum += road.getLength()
+
+        if not found_road:
+            raise AssertionError("Failed to pick a road. A distance beyound the last road must have been erroneously picked: {} (length sum: {}) (total length: {})".format(distance, length_sum, total_length))
```

Comment:
Assertions doesn't mesh with the logging, I found out but maybe we don't care about that.

Reply:
An assertion matches your description of critical, but it also halt's execution. If I an assertion is violated then code is defective, there for are there no reason to continue running.